Py 35 compat

MosesofEgypt opened the pull request:
Making changes in posix_compat compatible with python 3.5 and below, as well as fixing bugs and finishing up more of the posix compat work.

MosesofEgypt committed:
Tweaked various uses of pathlib.Path objects so they work below python 3.6(os.path functions don't accept Path objects below 3.6)
Made handlers be stored by their pathlib.Path tagsdir
Fixed bugs with animation and model compiler windows not properly determining tags dir when selecting tag.
Fixed missing imports.
Fixed bitmap conveter causing index out of range exception on log creation.
Refactoring code in app_window.

MosesofEgypt committed:
Made all tools, converters, and other mozz functionality work with posix compat(hopefully?)
Made several Mozzarilla properties decorated to always be a Path object.
Removed a tags_dir property from most tools, opting to use their handler.tagsdir property directly instead(less redundancy)

ShellyHerself commented on app_window.py:
```diff
@@ -153,7 +154,7 @@ def __init__(self, *args, **kwargs):
             pass
 
-        self.tags_dirs = [Path(e_c.WORKING_DIR, "tags")]
+        self.tags_dirs = [e_c.WORKING_DIR.joinpath("tags")]
```
```suggestion
        self.tags_dirs = [Path(tagpath_to_fullpath(e_c.WORKING_DIR, "tags", folder=True))]
```

ShellyHerself commented on app_window.py:
```diff
@@ -334,29 +335,56 @@ def __init__(self, *args, **kwargs):
     def tags_dir(self, new_val):
         assert isinstance(new_val, (str, Path))
-        self.tags_dirs[self._curr_tags_dir_index] = path_normalize(new_val)
+        self.tags_dirs[self._curr_tags_dir_index] = Path(path_normalize(new_val))
```
You're double dipping. Path already normalizes.
```suggestion
        self.tags_dirs[self._curr_tags_dir_index] = Path(new_val)
```

ShellyHerself commented on app_window.py:
```diff
@@ -334,29 +335,56 @@ def __init__(self, *args, **kwargs):
     def get_tags_dir_index(self, tags_dir):
         try:
-            return self.tags_dirs.index(path_normalize(tags_dir))
+            return self.tags_dirs.index(Path(path_normalize(tags_dir)))
```
```suggestion
            return self.tags_dirs.index(Path(tags_dir))
```

ShellyHerself commented on app_window.py:
```diff
@@ -383,7 +411,7 @@ def get_handler(self, index=None, tags_dir=None, create_if_not_exists=True):
                 index = self.handler_names.index(index)
 
-            tags_dir = path_normalize(tags_dir)
+            tags_dir = Path(path_normalize(tags_dir))
```
```suggestion
            tags_dir = Path(tags_dir)
```

ShellyHerself commented on app_window.py:
```diff
@@ -408,7 +436,7 @@ def create_handlers(self, tags_dir, handler_indices=()):
 
             handlers_by_dir = self.handlers[i]
-            if handlers_by_dir.get(tags_dir, None):
+            if handlers_by_dir.get(tags_dir):
```
Will this actually work if the path isn't a direct reference to that same path?

MosesofEgypt replied:
Good catch. We'll need to key handlers by their resolved tags dir, not a symlink or relative path. The equality operator only returns true(used in dict membership testing) for resolved paths.

```python
>>> from pathlib import Path
>>> asdf = Path("C:/Users/Moses/")
>>> qwer = Path("C:/Users/../Users/Moses/")
>>> qwer == asdf
False
>>> qwer.resolve() == asdf
True
```

ShellyHerself commented on app_window.py:
```diff
@@ -503,7 +532,7 @@ def add_tags_dir(self, e=None, tags_dir=None, manual=True):
             return
 
-        tags_dir = path_normalize(tags_dir)
+        tags_dir = Path(path_normalize(tags_dir))
```
```suggestion
        tags_dir = Path(tags_dir)
```

ShellyHerself commented on app_window.py:
```diff
@@ -553,7 +582,7 @@ def set_tags_dir(self, e=None, tags_dir=None, manual=True):
             return
 
-        tags_dir = path_normalize(tags_dir)
+        tags_dir = Path(path_normalize(tags_dir))
```
```suggestion
        tags_dir = Path(tags_dir)
```

ShellyHerself commented on widgets/directory_frame.py:
```diff
@@ -247,18 +250,20 @@ def activate_item(self, e=None):
 
         if os.path.isdir(tag_path):
-            app.last_load_dir = tag_path
+            app.last_load_dir = Path(tag_path)
```
```py
        tag_path = Path(tag_path)
        if tag_path.is_dir():
            app.last_load_dir = tag_path
```

ShellyHerself commented on windows/tag_converters/collision_converter.py:
```diff
@@ -60,7 +60,7 @@ def __init__(self, name):
 
 def get_tags(coll_path, model_in_path):
-    mod2_path = Path(str(Path(model_in_path).with_suffix('')) + "_COLLISION.gbxmodel")
+    mod2_path = str(Path(model_in_path).with_suffix('')) + "_COLL.gbxmodel"
```
Why are we using strings instead of paths for this class?

MosesofEgypt replied:
Simply for string concatenation. Path + str doesn't work in 3.5. Additionally, with the changes to supyr, a Tag's filepath property will always be converted to a Path when set, so we can leave it as a string here.

ShellyHerself commented on windows/tag_converters/converter_base.py:
```diff
@@ -38,6 +38,7 @@ def setup_window(self, *args, **kwargs):
             pass
 
+        # do path_replace to make sure the path works on linux
         tags_dir = path_replace(Path(curr_dir, 'tags'), 'tags', 'tags')
```
I've invented a solution that you should merge from posix compat in supyr.
```suggestion
         tags_dir = Path(path_split(curr_dir, "tags", after=True))
```

ShellyHerself commented on windows/tools/animations_compiler_window.py:
```diff
@@ -393,7 +394,7 @@ def jma_dir_browse(self):
 
         tags_dir = self.tags_dir.get()
-        data_dir = os.path.join(os.path.dirname(os.path.dirname(tags_dir)), "data", "")
+        data_dir = str(Path(tags_dir).parent.joinpath("data"))
```
We still need a better case insensitive solution to this.

ShellyHerself replied:
```suggestion
        # Add data to the path and then use path_replace to match the case of any
        # data directory that might already be here. 
        data_dir = str(path_replace(Path(tags_dir).parent.joinpath("data"), "data", "data"))
```

ShellyHerself commented on windows/tools/animations_compiler_window.py:
```diff
@@ -430,7 +435,7 @@ def tags_dir_browse(self):
             return
 
-        tags_dir = path_normalize(tags_dir)
+        tags_dir = str(Path(path_normalize(tags_dir)))
```
```suggestion
        tags_dir = str(Path(tags_dir))
```

ShellyHerself commented on windows/tools/animations_compiler_window.py:
```diff
@@ -456,14 +461,15 @@ def model_animations_path_browse(self, force=False):
+        self.tags_dir.set(
+            os.path.join(
+                path_split(self.app_root.last_load_dir, "tags"),
+                "tags"))
```
See earlier comment about the after arg for path_split

ShellyHerself replied:
```py
        self.tags_dir.set(path_split(self.app_root.last_load_dir, "tags", after=True))
```

ShellyHerself commented on windows/tools/animations_compression_window.py:
```diff
@@ -340,7 +342,8 @@ def _do_compression(self, compress, antr_path=None):
                 fp = Path(antr_tag.filepath)
-                antr_tag.filepath = Path(fp.parent, fp.stem+"_DECOMP"+fp.suffix)
+                antr_tag.filepath = fp.parent.joinpath(
+                    "%s_DECOMP%s" % (fp.stem, fp.suffix))
```
This change should be undone for readability imo.

MosesofEgypt replied:
Agreed, but I will be putting spaces between the + operators. If we're talking readability, there needs to be space between operators in a statement.

ShellyHerself commented on windows/tools/bitmap_converter_window.py:
```diff
@@ -862,17 +863,16 @@ def _scan(self):
                     # bitmap tag info as hash keys. This is just so that we
                     # don't do any double conversions.
-                    fp = str(Path(rel_root, filename).resolve())
+                    fp = str(rel_root.joinpath(filename))
```
This doesn't line up with the comment above now, I wouldn't do this.

MosesofEgypt replied:
Path.resolve doesn't work on 3.5 if it can't be resolved. There is no optional "strict" parameter to the resolve method. Also, it wasn't being stored as an absolute path before, just a partially resolved path. For it to be absolute, it would have had to be joined with root.

I have implemented a set to keep track of all absolute paths(joined with root) as the scan progresses to make sure no double conversions happen. How's that?

ShellyHerself commented on windows/tools/bitmap_converter_window.py:
```diff
@@ -1277,7 +1277,7 @@ def make_log(self):
             tag_info_strs[typ]  = [''] * 18
 
-            for fmt in range(len(BITMAP_FORMATS)):
+            for fmt in range(min(len(BITMAP_FORMATS), 18)):
```
Idk what you're doing here. It's out of scope of this pr, but if you're going to do it, explain it.

MosesofEgypt replied:
Due to supporting halo 2 and halo 3, I ended up implementing more bitmap formats than our 18. Because BITMAP_FORMATS is pulled from the globally defined halo bitmap formats enum, there will be more than 18 in it, thus causing an IndexError when it tries to access formatted_strs[typ][18]

A proper fix would be to define a constant at the top of the file for the max number of halo 1 formats, and work from that like so.
```python

HALO_1_TYPE_COUNT   = 4
HALO_1_FORMAT_COUNT = 18
for typ in range(HALO_1_TYPE_COUNT):
    formatted_strs[typ] = [''] * HALO_1_FORMAT_COUNT
    tag_info_strs[typ]  = [''] * HALO_1_FORMAT_COUNT

    for fmt in range(HALO_1_FORMAT_COUNT):
        if "?" not in BITMAP_FORMATS[fmt]:
            formatted_strs[typ][fmt] = "\n\n\t%s" % BITMAP_FORMATS[fmt]
            tag_info_strs[typ][fmt] = {}
```

I am fixing this now because it was preventing me from testing other changes, as it would always exception when making a log.

ShellyHerself commented on windows/tools/bitmap_source_extractor_window.py:
```diff
@@ -41,8 +41,8 @@ def __init__(self, app_root, *args, **kwargs):
-        self.tags_dir.set(os.path.join(e_c.WORKING_DIR, 'tags'))
-        self.data_dir.set(os.path.join(e_c.WORKING_DIR, 'data'))
+        self.tags_dir.set(e_c.WORKING_DIR.joinpath('tags'))
+        self.data_dir.set(e_c.WORKING_DIR.joinpath('data'))
```
Why not `Path(e_c.WORKING_DIR, "tags")` it's shorter and does the same thing.

MosesofEgypt replied:
Because the word "joinpath" makes it more apparent what is being done without knowing how the Path constructor works. Yes, it's slightly longer, but this is the intended usage of joinpath. It also ensures that the subclass of Path will be the same as the Path whose joinpath method was invoked.

ShellyHerself replied:
If someone doesn't know what Path is they should look it up. And the subclass doesn't matter.

ShellyHerself commented on windows/tools/compile_bitmap.py:
```diff
@@ -127,8 +129,7 @@ def bitmap_from_bitmap_source(app, e=None):
         window.tag.rel_filepath = "untitled%s.bitmap" % app.untitled_num
-        window.tag.filepath = os.path.join(
-            app.tags_dir + window.tag.rel_filepath)
+        window.tag.filepath = app.tags_dir.joinpath(window.tag.rel_filepath)
```
Please also just go for `Path(a, b)` here instead of `a.joinpath(b)`

MosesofEgypt replied:
See above comment.

ShellyHerself commented on windows/tools/compile_hud_message_text.py:
```diff
@@ -1,66 +1,69 @@
-        tag_path = os.path.join(tags_dir, rel_tagpath)
-    else:
-        tag_path = ""
+    rel_filepath = rel_filepath.joinpath("hud messages.hud_message_text")
```
Path(a, b)

MosesofEgypt replied:
See above comment.

ShellyHerself commented on windows/tools/compile_hud_message_text.py:
```diff
@@ -1,66 +1,69 @@
-        tag_load_path = ""
+    tag_path = Path("")
+    if not is_path_empty(rel_filepath):
+        tag_path = tags_dir.joinpath(rel_filepath)
```
I will pick all the nits

MosesofEgypt replied:
See above comment.

ShellyHerself commented on windows/tools/compile_physics.py:
```diff
@@ -1,75 +1,72 @@
+    
+    try:
+        rel_filepath = fp.relative_to(data_dir).parent.parent
+        rel_filepath = rel_filepath.joinpath("%s.physics" % rel_filepath.stem)
```
.with_suffix()

MosesofEgypt replied:
with_suffix raises an exception if the suffix isn't empty and doesn't start with a period. Unless I'm missing your point.

ShellyHerself replied:
Suffix has no crash condition for when there already is a suffix.
https://docs.python.org/3.5/library/pathlib.html#pathlib.PurePath.with_suffix
```suggestion
        rel_filepath = rel_filepath.with_suffix(".physics")
```
Also, what you wanted to do you needed the name for.
https://docs.python.org/3.5/library/pathlib.html#pathlib.PurePath.name

This code right now would turn `this/lovely/long/path.ext` into `this/lovely/long/path/this/lovely/long/path.physics`

ShellyHerself commented on windows/tools/dependency_window.py:
```diff
@@ -204,11 +199,9 @@ def do_recursive_zip(self):
 
-        tag_path = path_normalize(tag_path)
-        if not is_in_dir(tag_path, tags_dir):
+        tag_path = Path(path_normalize(tag_path))
```
Double dip

ShellyHerself commented on windows/tools/model_compiler_window.py:
```diff
@@ -407,6 +407,7 @@ def jms_dir_browse(self):
         if not dirpath:
             return
 
+        dirpath = str(Path(dirpath))
         if tags_dir and data_dir and os.path.basename(dirpath).lower() == "models":
             object_dir = os.path.dirname(dirpath)
```
Please make this whole segment use the Path functions.

ShellyHerself commented on windows/tools/model_compiler_window.py:
```diff
@@ -434,6 +435,8 @@ def tags_dir_browse(self):
         if not tags_dir:
             return
 
+        tags_dir = str(Path(path_normalize(tags_dir)))
```
Triple dip. If you need it as a string, just do str(path_normalize

ShellyHerself commented on windows/tools/model_compiler_window.py:
```diff
@@ -460,12 +463,13 @@ def gbxmodel_path_browse(self, force=False):
-        )
+            os.path.join(
+                path_split(self.app_root.last_load_dir, "tags"),
+                "tags"))
```
path_split with after=True

ShellyHerself commented on windows/tools/tag_scanner_window.py:
```diff
@@ -283,10 +280,10 @@ def scan(self):
 
             for filename in files:
-                filepath = os.path.join(rel_root, path_normalize(filename))
+                filepath = rel_root.joinpath(path_normalize(filename))
```
Little dipper

ShellyHerself commented on windows/tools/tag_scanner_window.py:
```diff
@@ -321,10 +318,10 @@ def scan(self):
                 if time() - c_time > p_int:
                     c_time = time()
-                    print(' '*4 + filepath)
+                    print(' '*4, filepath)
```
That's indented with 5 spaces now.

ShellyHerself commented on windows/tools/tag_scanner_window.py:
```diff
@@ -480,5 +477,5 @@ def tag_specific_scan(self, tag, errors):
             errors[cls] = "%s\n%s:\n%s\n" % (
                 errors.get(cls, ""),
-                tag.filepath.split(self.tags_dir, 1)[-1],
+                str(tag.filepath).split(str(self.handler.tagsdir), 1)[-1],
```
What actually is this code?

MosesofEgypt replied:
Piecing together a tag-specific error string. I have cleaned it up to make it more clear.
```python
rel_tag_path = str(tag.filepath.relative_to(self.handler.tagsdir))
errors[cls] = (
    errors.get(cls, "") + "\n" +
    rel_tag_path + "\n" +
    err + "\n"
    )
```

ShellyHerself reviewed (changes requested):
Welcome, to posix compat.

MosesofEgypt committed:
Implementing some of the requested changes in pull request.
Implemented storing tags dirs(and handlers by their tags dirs) as resolved paths to ensure proper set comparison(symlinks and relative paths won't be equal until resolved).

MosesofEgypt committed:
Implementing more changes laid out in pull request comments.

ShellyHerself commented on app_window.py:
```diff
@@ -73,12 +72,19 @@ class Mozzarilla(Binilla):
+    tags_dirs = ()
+
+    _styles_dir  = Path(e_c.SETTINGS_DIR, "styles")
+    _config_path = Path(e_c.SETTINGS_DIR, "mek_config", "mozzarilla.cfg")
```
No! The settings dir in editor_constants should be set to be WORKING_DIR, "mek_config" on windows. Not like this.

ShellyHerself replied:
I'd compare it to what the Posix version looks like in that same file. Make something similar, but in your style.

ShellyHerself commented on app_window.py:
```diff
@@ -334,29 +340,68 @@ def __init__(self, *args, **kwargs):
+        try:
+            new_val = new_val.resolve()
+        except FileNotFoundError:
+            pass
```
DON'T RESOLVE. This fucks with people's symlink setups. There is no reason to do it here.

We only used resolve in the bitmap converter to check if two files are the same.

ShellyHerself commented on app_window.py:
```diff
@@ -334,29 +340,68 @@ def __init__(self, *args, **kwargs):
+            try:
+                tags_dir = tags_dir.resolve()
+            except FileNotFoundError:
+                pass
```
NO!

ShellyHerself commented on app_window.py:
```diff
@@ -383,20 +428,30 @@ def get_handler(self, index=None, tags_dir=None, create_if_not_exists=True):
+            try:
+                full_tags_dir = tags_dir.resolve()
+            except FileNotFoundError:
+                full_tags_dir = tags_dir
```
No.

ShellyHerself commented on app_window.py:
```diff
@@ -383,20 +428,30 @@ def get_handler(self, index=None, tags_dir=None, create_if_not_exists=True):
+        try:
+            full_tags_dir = tags_dir.resolve()
+        except FileNotFoundError:
+            full_tags_dir = tags_dir
```
Seriously, we actually cannot do this if we want the tools to behave as expected.

ShellyHerself commented on app_window.py:
```diff
@@ -503,13 +557,17 @@ def add_tags_dir(self, e=None, tags_dir=None, manual=True):
+        try:
+            full_tags_dir = tags_dir.resolve()
+        except FileNotFoundError:
+            full_tags_dir = tags_dir
```
![](https://images-cdn.9gag.com/photo/aAD5ByZ_460s.jpg)

ShellyHerself commented on app_window.py:
```diff
@@ -553,8 +611,13 @@ def set_tags_dir(self, e=None, tags_dir=None, manual=True):
+        try:
+            full_tags_dir = tags_dir.resolve()
+        except FileNotFoundError:
+            full_tags_dir = tags_dir
```
![](https://external-preview.redd.it/7GiN7IFQGB3WjLJb9UEiwUR4kaq6KLdLtGSSoEHAqFQ.jpg?auto=webp&s=e7300b2fe5c79170e5c5e63e2819ccdd1ee3b838)

ShellyHerself commented on windows/tools/animations_compiler_window.py:
```diff
@@ -416,7 +418,10 @@ def jma_dir_browse(self):
         self.jma_dir.set(dirpath)
         if not self.tags_dir.get():
-            self.tags_dir.set(path_replace(self.app_root.last_load_dir, "data", "tags"))
+            self.tags_dir.set(
+                os.path.join(
+                    path_split(self.app_root.last_load_dir, "data"),
+                    "tags"))
```
The handler class needs its own data dir property. We can't keep making up different ways to do this every time we meet this problem.

ShellyHerself commented on windows/tools/animations_compiler_window.py:
```diff
@@ -456,14 +461,15 @@ def model_animations_path_browse(self, force=False):
-        fp = Path(fp)
-        if not os.path.splitext(fp)[-1]:
-            fp += ".model_animations"
+        fp = Path(fp).with_suffix(".model_animations")
```
This will fuck up on Posix if there is any capitals in the original fp's file extension.

ShellyHerself commented on windows/tools/bitmap_converter_window.py:
```diff
@@ -52,6 +53,9 @@
 PARAM_FORMAT_TO_FORMAT = (-1, ) + VALID_FORMAT_ENUMS
 FORMAT_OPTIONS = ("Unchanged", ) + tuple(BITMAP_FORMATS[i] for i in VALID_FORMAT_ENUMS)
 
+HALO_1_TYPE_COUNT   = 4
+HALO_1_FORMAT_COUNT = 18
```
```
HALO_1_BITMAP_TYPE_COUNT
HALO_1_BITMAP_FORMAT_COUNT
```

ShellyHerself committed:
[CodeReview] Apply some suggestions directly

ShellyHerself committed:
[CodeReview] Implement requested changes from pullrequest

ShellyHerself committed:
[CodeReview] Fix small mistake

ShellyHerself committed:
[CodeReview] Undo renaming of config and styles paths.
Code was resorting to basing its config paths off of the ones found in 
Binilla.
The feature was basically only half implemented. >:|

ShellyHerself committed:
[app_window] Redo config path var rename

ShellyHerself committed:
[widgets] Fix rawdata widgets not being able to export sound

MosesofEgypt committed:
Fixed bug with hud_message compilation where the wrong filepath was generated for the resulting tag.
Removed debug code from compile_strings

MosesofEgypt committed:
Changed how custom iconbitmap is applied for tag windows
Created custom config class with custom iconbitmap filepath
Fixed bug with converter_base not importing path_split

ShellyHerself commented on windows/tag_window.py:
```diff
@@ -31,3 +39,12 @@ def use_scenario_names_for_script_names(self):
+
+class HaloConfigWindow(ConfigWindow):
+
+    def post_toplevel_init(self):
+        if not is_path_empty(e_c.MOZZ_ICON_PATH):
+            self.iconbitmap_filepath = e_c.MOZZ_ICON_PATH
```
Don't we need a try-except here to avoid this process exceptioning under Linux?

MosesofEgypt replied:
why would that exception? is that property not defined on linux?

ShellyHerself replied:
Linux tkinter not supporting directly assigning icon paths as Window icons. It's the whole reason for all of those try except blocks or did I understand that wrong?

MosesofEgypt commented on editor_constants.py:
```diff
@@ -39,7 +39,7 @@
-    SETTINGS_DIR = WORKING_DIR
+    SETTINGS_DIR = Path(WORKING_DIR, "mek_config")
 else:
     SETTINGS_DIR = Path(Path.home(), ".local", "share", "mek", "mozzarilla")
```
The SETTINGS_DIR on non-windows systems should mirror the windows one, and be Path(Path.home(), ".local", "share", "mek_config")

MosesofEgypt commented on widgets/field_widgets/rawdata_frames.py:
```diff
@@ -180,6 +180,8 @@ def export_node(self):
             filepath = filepath.with_suffix(def_ext)
 
+        ext = filepath.suffix
+
```
ext is being compared to lowercase strings. Need to lowercase it

ShellyHerself committed:
[e_c] Remove mozzarilla sub dir from settings fir constant

ShellyHerself committed:
[widgets] case-insentively compare extension in export_node

MosesofEgypt committed:
Modified how WORKING_DIR and MOZZLIB_DIR are defined

MosesofEgypt committed:
Don't use with_suffix when it can't be guaranteed there is a suffix. Use string concatenation instead.

ShellyHerself committed:
[e_c] Fix compatibility for some path constants
